### ISSUE TITLE
Forcing one-time run mode for docker

### ISSUE THREAD

**Issue by 2fst4u** [recyclarr/recyclarr#103]
In the wiki the instructions outline the difference run methods being docker run and Docker compose. The former being a one-time run and the latter being a Cron by default.

I run my recyclarr on kubernetes, so these methods aren't exactly analogous and the container is always running, so it's running in from mode.

It would be good if I could force the one-time run mode with an environment variable, or alternatively (and potentially less popular), make it one-time run by default _unless_ a Cron is specific.

This way you would be able to utilise the kubernetes cron method of deployment which means the container resources aren't scheduled until the script needs to run, and then it's shut down. Basically like a scheduled docker run command.

**Comment by skayo** [comment]
The container will do a one-time run if any arguments are passed to the container (`CMD` field in Dockerfile), otherwise it will run a cron daemon:
```
docker compose run --rm recyclarr sonarr   <-- One time (syncs sonarr and exists)
docker compose run --rm recyclarr          <-- Starts cron daemon (will never exit by itself)
```
So you could create a kubernetes deployment that just runs `recyclarr sonarr` and `recyclarr radarr` on a specific schedule.
But I'm not at all familiar with kubernetes, so we might need another solution for this.

**Comment by 2fst4u** [comment]
Yea so this doesn't really work for kubernetes because a pod appears to run basically like a docker-compose up command. I have a deployment of recyclarr running full-time with no arguments passed to it and it runs daily like the compose method does.

Hence the desire to just force it with a flag. Giving that control would just give peace of mind.

**Comment by rcdailey** [comment]
You haven't really explained what problem you're trying to solve. Why do you need to do a 1 time run in your kubernetes setup? Why not just run it locally on a machine somewhere? What's your use case? Once you explain your situation, requirements, why you're doing things, etc, we can go from there. It's possible there's a different solution to your problem.

Helpful information at the link below if you're willing to read. I am not trying to be rude by sending you this. Someone sent this to me years ago and it really changed my view on how to properly ask for help. Hopefully it helps you too. https://xyproblem.info/

**Comment by 2fst4u** [comment]
I did explain why.

> 
> This way you would be able to utilise the kubernetes cron method of deployment **which means the container resources aren't scheduled until the script needs to run**, and then it's shut down. Basically like a scheduled docker run command.

To save idle resources.

**Comment by rcdailey** [comment]
I don't really understand what that means. I do not use kubernetes. You'll need to dumb it down.

I assumed you would be requesting this for some specific use case, but all you're looking for is (what sounds like) an optimization? Why do "idle resources" matter? Again, it feels like this is an out of context request. I'll assume for the moment though that it's just that I don't understand enough about Kubernetes.

I'll wait for you to explain more detail as I originally requested. Thank you!

**Comment by 2fst4u** [comment]
There's probably a few kubernetes concepts that could be explained here but I guess the main one is scheduling.

In kubernetes you can just set containers to run without any resource reservations or restrictions and let them use whatever they can, but this causes contention issues if you don't manage it well.

So generally people use CPU and memory "requests" and "limits", with requests being the important part.

A request is a guarantee that those resources will be available to your containers, which in turn means it's set aside and not able to be used by other containers you have running.

So to make sure recyclarr runs properly and doesn't get any OOM errors, or just has sufficient CPU cycles to run, I am using requests to guarantee these resources to recyclarr.

This means those requested resources aren't available to my other containers which could be using it the other 23 hours of the day when recyclarr isn't running.

So kubernetes has a cronjob feature, where pods can be scheduled on a cronjob and then when the container completes its task and exits, the container is deleted and the job is complete.

This means those resource reservations are still used, but only for the period of time the container is running. Then when it is no longer running those scheduled resources are freed up for other pods and containers to use.

I use this method for plex-meta-manager and it works great. That project can use quite a lot of resources if you library and config get quite large, so the resources getting freed up is useful.

While recyclarr doesn't use all that much resources, it would still be good for the same reasons, especially because I run all these on a raspberry pi cluster so resources are a premium.

Making recyclarr force a one-time run and exit the script would enable this usage of kubernetes cronjobs rather than having the container run 24/7 not using the reserved resources most of the time.

I hope that made enough sense.

**Comment by rcdailey** [comment]
Thanks for the detailed explanation. I think all that makes sense.

Is K8S not able to perform the equivalent of a `docker run --rm`? Or put another way, when you tell it to run a container, can you not pass it arguments?

**Comment by 2fst4u** [comment]
There is but that's a bit antithetical and at that point it's not being automated. I may as well just run the script on a local machine if I were to do that and I want to automate the process entirely.

**Comment by rcdailey** [comment]
Antithetical to what? What kind of automation are you looking for? In my mind, putting the container in cron mode to run on a schedule is how you automate it. What other kind of behavior do you need? I understand you want the container to be instructed to run once and then exit, but in what way will you automate this?

**Comment by rcdailey** [comment]
Just to make sure I'm not going in circles on this, let's try an example. As I said, I'm not familiar with Kubernetes, so I took a sample I found on stackoverflow for a cron job and I'm modifying it for example purposes. This YAML may be incorrect but hopefully it gets some more helpful discussion going.

```yml
apiVersion: batch/v1beta1
kind: CronJob
metadata:
  name: app
spec:
  schedule: "* * * * *"
  jobTemplate:
    spec:
      template:
        spec:
          containers:
          - name: recyclarr
            image: ghcr.io/recyclarr/recyclarr
            args: ["sonarr"]
          restartPolicy: OnFailure
```

Would something like this not work?

**Comment by 2fst4u** [comment]
> I understand you want the container to be instructed to run once and then exit, but in what way will you automate this?

Using kubernetes cronjobs. I feel like I've been through this.

https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/

Kubernetes schedules the container in a pod. Script runs. Pod is broken down and no longer impacts scheduling.

No hands on input or intervention required and the script runs automatically on a schedule set by kubernetes, rather than a schedule set internally by the script as this requires the pod to be running 24/7.




> Just to make sure I'm not going in circles on this, let's try an example. As I said, I'm not familiar with Kubernetes, so I took a sample I found on stackoverflow for a cron job and I'm modifying it for example purposes. This YAML may be incorrect but hopefully it gets some more helpful discussion going.
> 
> ```yml
> apiVersion: batch/v1beta1
> kind: CronJob
> metadata:
>   name: app
> spec:
>   schedule: "* * * * *"
>   jobTemplate:
>     spec:
>       template:
>         spec:
>           containers:
>           - name: recyclarr
>             image: ghcr.io/recyclarr/recyclarr
>             args: ["sonarr"]
>           restartPolicy: OnFailure
> ```
> 
> Would something like this not work?

No, in this example kubernetes would schedule _one whole brand new pod_ every single minute of every day for eternity. That's what `schedule: "* * * * *"` means. This would mean 1,440 brand new pods would attempt to schedule each day.

A more logical use-case would be the same thing but with `schedule: "0 0 * * *"` for example, as this would schedule a _single pod_ at midnight, wait for the pod to complete its task, then exit and only do this once per day.

But currently this wouldn't be possible with how recyclarr is set up by default as kubernetes would schedule the pod at this time, then recyclarr would internally wait on its own cron schedule inside the pod, not do anything, and the pod would be shut down. 

The cron scheduling of kubernetes and the internal script cron are unrelated and don't interact at all. This would cause a nested cron task essentially, but with the kubernetes one not caring about the other and just shutting down the pod prematurely.

**Comment by rcdailey** [comment]
You're missing the `args` part of the example I gave, which is the important bit. Forget how often the job runs in the example. That's not important or relevant to the topic here. If you provide args, the container runs once and then exits. It's semantically identical to `pmm --run` in Plex Meta Manager.

**Comment by 2fst4u** [comment]
Aaah I see. That's potentially the key part that wasn't translating over to the way I understood the container ran. I could never figure out how that part of the documentation forced it to run once instead of cron mode.

I'll try that and see how it goes.
